src/coreclr/jit/async.cpp

```diff
@@ -0,0 +1,1977 @@
+//   defs       - Current outstanding LIR edges
+//   liveLocals - [out] Vector to add new live local information into
+//
+void AsyncTransformation::LiftLIREdges(BasicBlock*                     block,
+                                       const jitstd::vector<GenTree*>& defs,
+                                       jitstd::vector<LiveLocalInfo>&  liveLocals)
+{
+    if (defs.size() <= 0)
+    {
+        return;
+    }
+
+    for (GenTree* tree : defs)
+    {
+        // TODO-CQ: Enable this. It currently breaks our recognition of how the
+        // call is stored.
+        // if (tree->OperIs(GT_LCL_VAR))
+        //{
+        //    LclVarDsc* dsc = m_comp->lvaGetDesc(tree->AsLclVarCommon());
+        //    if (!dsc->IsAddressExposed())
+        //    {
+        //        // No interference by IR invariants.
+        //        LIR::AsRange(block).Remove(tree);
+        //        LIR::AsRange(block).InsertAfter(beyond, tree);
+        //        continue;
+        //    }
+        //}
+
+        LIR::Use use;
+        bool     gotUse = LIR::AsRange(block).TryGetUse(tree, &use);
+        assert(gotUse); // Defs list should not contain unused values.
+
+        unsigned newLclNum = use.ReplaceWithLclVar(m_comp);
+        liveLocals.push_back(LiveLocalInfo(newLclNum));
+        GenTree* newUse = use.Def();
+        LIR::AsRange(block).Remove(newUse);
+        LIR::AsRange(block).InsertBefore(use.User(), newUse);
+    }
+}
+
+//------------------------------------------------------------------------
+// AsyncTransformation::LayOutContinuation:
+//   Create the layout of the GC pointer and data arrays in the continuation
+//   object.
+//
+// Parameters:
+//   block      - The block containing the async call
+//   call       - The async call
+//   liveLocals - [in, out] Information about each live local. Size/alignment
+//                information is read and offset/index information is written.
+//
+// Returns:
+//   Layout information.
+//
+ContinuationLayout AsyncTransformation::LayOutContinuation(BasicBlock*                    block,
+                                                           GenTreeCall*                   call,
+                                                           jitstd::vector<LiveLocalInfo>& liveLocals)
+{
+    ContinuationLayout layout(liveLocals);
+
+    for (LiveLocalInfo& inf : liveLocals)
+    {
+        LclVarDsc* dsc = m_comp->lvaGetDesc(inf.LclNum);
+
+        if ((dsc->TypeGet() == TYP_STRUCT) || dsc->IsImplicitByRef())
+        {
+            ClassLayout* layout = dsc->GetLayout();
+            assert(!layout->HasGCByRef());
+
+            if (layout->IsCustomLayout())
+            {
+                inf.Alignment   = 1;
+                inf.DataSize    = layout->GetSize();
+                inf.GCDataCount = layout->GetGCPtrCount();
+            }
+            else
+            {
+                inf.Alignment = m_comp->info.compCompHnd->getClassAlignmentRequirement(layout->GetClassHandle());
```

Comment:
Does alignment actually matter here?

Reply:
Only for arm32 doubles/floats, probably. We could mark the loads in the restore path as unaligned instead.